ALC Loading interface to MVP

GuiMacielPereira opened the pull request:
### Description of work
Changed code for ALC Loading interface to follow the MVP architecture.

#### Summary of work
<!-- Please provide a short, high level description of the work that was done.
-->
- Created abstract class for presenter
- Made presenter NOT a QObject
- Put all Qt code inside the view
- Created new class for model
- Created abstract class for model
- Made abstract class of view NOT a QObject
- Corrected presenter unit test


<!-- Why has this work been done? If there is no linked issue please provide appropriate context for this work.
#### Purpose of work
This can be removed if a github issue is referenced below
-->
Followed the requests stated in #38009 .

Fixes #38009 . <!-- and fix #xxxx or close #xxxx xor resolves #xxxx. One line per issue fixed. -->
<!-- alternative
*There is no associated issue.*
-->

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

#### Further detail of work
<!-- Please provide a more detailed description of the work that has been undertaken.
-->

### To test:
- Code review
- Manual testing (Need to have directory with training dataset):
  - Open Muon > ALC interface
  - Change instrument to MUSR
  - For the runs, type 15189
  - You should see some green text indicating the run has been found
  - Click in manage user directories, check that interface opens
  - Click in Period info, check that it opens
  - Click Load
  - After loading, click Auto add
  - Go to the local directory where the file is being loaded from and copy the `MUSR15189` file and rename the copy to `MUSR15190`
  - You should see the interface loads the file and the runs text box is now updated with the run 15189
  - Go to the data modelling page
  - You should see the same data displayed as in the data loading interface

<!-- Instructions for testing.
There should be sufficient instructions for someone unfamiliar with the application to test - unless a specific
reviewer is requested.
If instructions for replicating the fault are contained in the linked issue then it is OK to refer back to these.
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

GuiMacielPereira committed:
Put presenter calls inside view slots

Instead of the previous calls:
view -> emit signal -> presenter catches signal -> presenter slot
Now we have:
view -> emit signal -> view slot -> presenter handle

This makes the code simpler and avoids the presenter being a QObject

GuiMacielPereira committed:
Replace signal by calling presenter directly

Option to toggle auto add would trigger a signal that would call the
function startWatching in the presenter.
Replaced this by directly calling the function from the view, since now
the presenter is accessible from the view.

GuiMacielPereira committed:
Remove signal from presenter

Signal dataChanged() from presenter was changed to signal emmited from the view.

GuiMacielPereira committed:
Make presenter a non QObject

Removed inheritance of presenter as a QObject.
Instead of using QObject::timerEvent(), replaced it by using QTimer.
Replaced connect() with QObject::connect() and replaced QObject slots with lambdas,
which is possible since qt5.

GuiMacielPereira committed:
Removed unused slots and signals

Cleaned slots and signals that were now unused due to work from previous commits.
Realised that one of the slots called instrumentChanged() should be used instead of
creating a new one that notifies the presenter. Corrected this oversight.

GuiMacielPereira committed:
Create abstract class IALCDataLoadingPresenter

Created virtual methods for the public methods in the ALCDataLoadingPresenter class.
Changed arguments to use the abstract class.

GuiMacielPereira committed:
Add destructor to virtual class

GuiMacielPereira committed:
Pass opening of manage directories to view

Because it contains qt code and is very straight forward.

GuiMacielPereira committed:
Move watcher and timer into view

Previuosly the watcher and timer in the presenter were QObjects.
To make the presenter not dependent on qt I moved them into the view.

GuiMacielPereira committed:
Move period info interface to view

View now owns the period info interface, so that presenter is not dependent on qt anymore.

GuiMacielPereira committed:
Moved watcher and timer logic to presenter

Realised that logic could be moved to presenter by introducing getter
methods from the view.

GuiMacielPereira committed:
Create model

Added source and header files for model.
Passed load() function to model since that is where a lot of the calculations happen.

GuiMacielPereira committed:
Pass some variables to model fully

Put code to do with some variables that store the state whithin model fully and deleted those variables from the presenter

GuiMacielPereira committed:
Pass updateAvailableInfo() to model

Put all of the logic to update information on logs and periods into the model.

GuiMacielPereira committed:
Pass extractRunNumber() into model

GuiMacielPereira committed:
Pass getPathFromFiles to model

GuiMacielPereira committed:
Pass checkGroupingValid() to model

GuiMacielPereira committed:
Finish separation between model and presenter

This commit finishes the separation between presenter and model, by passing the function
timerEvent() to the model. Now the presenter doesn't have any state variables and holds only
the view and the model.

GuiMacielPereira committed:
Change variable name and less arguments for load()

GuiMacielPereira committed:
Fix seg fault crash

Previously the ALC interface was crashing Mantid when it was closed.
This was due to two pointers (m_watcher and m_timer) being deleted twice.
I am not sure what was happening but I reckon that since they were owned by
the loading interface, upon closing the ALC interface they were being deleted by
the loading interface and the ALC (parent) interface both, causing a double deletion error.
I fixed it by passing ownership of the wather and timere to the ALC (parent) interface.

GuiMacielPereira committed:
Fix cppcheck errors

GuiMacielPereira committed:
Fix unmatched declaration

GuiMacielPereira committed:
Clean header file and function declarations

Added const and static to function declarations.

GuiMacielPereira committed:
Add abstract class for model

Added abstract class for model, changed the constructor of presenter to include
this abstract class and cleaned ALCInterface.cpp where the presenter is first initiated.

GuiMacielPereira committed:
Correct unit test of presenter

Modified file for testing presenter according to the changes I made before.

GuiMacielPereira committed:
Fix cpp check error

Was missing an override for the model destructor

GuiMacielPereira committed:
Make view inherit QObject instead of abstract parent

Instead of abstract class of view inheriting qobject,
made class of view a qobject. Simplifies signals and slots.

robertapplin commented on qt/scientific_interfaces/Muon/test/ALCDataLoadingPresenterTest.h:
```diff
@@ -131,10 +148,15 @@ class ALCDataLoadingPresenterTest : public CxxTest::TestSuite {
   void setUp() override {
-    m_view = new NiceMock<MockALCDataLoadingView>();
-    m_presenter = new ALCDataLoadingPresenter(m_view);
+    m_view = std::make_unique<NiceMock<MockALCDataLoadingView>>();
+    auto model = std::make_unique<ALCDataLoadingModel>();
+    m_presenter = std::make_unique<ALCDataLoadingPresenter>(m_view.get(), std::move(model));
```
It would be useful to keep a member variable which is a non-owning raw pointer to the model. This means that we will be able to assert expectations on the model in the below unit tests by using the `m_model` member variable. (See the 3rd green line below)
```suggestion
    m_view = std::make_unique<NiceMock<MockALCDataLoadingView>>();
    auto model = std::make_unique<ALCDataLoadingModel>();
    m_model = model.get();
    m_presenter = std::make_unique<ALCDataLoadingPresenter>(m_view.get(), std::move(model));
```

GuiMacielPereira replied:
I see!

robertapplin commented on qt/scientific_interfaces/Muon/IALCDataLoadingModel.h:
```diff
@@ -0,0 +1,47 @@
+  virtual void updateAutoLoadCancelled() = 0;
+  virtual bool loadFilesFromWatchingDirectory(const std::string &firstFile, const std::vector<std::string> &files,
+                                              const std::string &runsText) = 0;
+  virtual std::string getPathFromFiles(std::vector<std::string> files) = 0;
```
Can this parameter be `const &`, and the function be `const`?
```suggestion
  virtual std::string getPathFromFiles(const std::vector<std::string> &files) const = 0;
```

robertapplin commented on qt/scientific_interfaces/Muon/IALCDataLoadingModel.h:
```diff
@@ -0,0 +1,47 @@
+  virtual std::string getPathFromFiles(std::vector<std::string> files) = 0;
+
+  // Getters
+  virtual bool getLoadingData() = 0;
+  virtual Mantid::API::MatrixWorkspace_sptr getLoadedData() = 0;
+  virtual std::vector<std::string> &getLogs() = 0;
+  virtual std::vector<std::string> &getPeriods() = 0;
+  virtual Mantid::API::MatrixWorkspace_sptr getWsForMuonInfo() = 0;
+  virtual double getMinTime() const = 0;
+  virtual std::string &getRunsText() = 0;
```
Can any of these getters be `const`?

GuiMacielPereira replied:
Well spotted, all of them can be const, except for the ones returning by reference.

robertapplin commented on qt/scientific_interfaces/Muon/ALCDataLoadingView.cpp:
```diff
@@ -76,6 +72,24 @@ void ALCDataLoadingView::initialize() {
   m_ui.runs->doButtonOpt(MantidQt::API::FileFinderWidget::ButtonOpts::None);
+
+  // Set up connections
+  connect(m_ui.help, SIGNAL(clicked()), this, SLOT(help()));
+  connect(m_ui.load, SIGNAL(clicked()), this, SLOT(notifyLoadClicked()));
+  connect(m_ui.instrument, SIGNAL(currentTextChanged(QString)), this, SLOT(instrumentChanged(QString)));
+  connect(m_ui.runs, SIGNAL(fileTextChanged(const QString &)), this, SLOT(notifyRunsEditingChanged()));
+  connect(m_ui.runs, SIGNAL(findingFiles()), this, SLOT(notifyRunsEditingFinished()));
+  connect(m_ui.runs, SIGNAL(fileFindingFinished()), this, SLOT(notifyRunsFoundFinished()));
+  connect(m_ui.manageDirectoriesButton, SIGNAL(clicked()), this, SLOT(openManageDirectories()));
+  connect(m_ui.periodInfo, SIGNAL(clicked()), this, SLOT(notifyPeriodInfoClicked()));
+  connect(m_ui.runsAutoAdd, SIGNAL(toggled(bool)), this, SLOT(runsAutoAddToggled(bool)));
+
+  // Watcher to check if directory to load data from changes
+  // Need to connect using function pointers because directoryChanged() is somehow not recognised as a signal
+  connect(m_watcher, &QFileSystemWatcher::directoryChanged, [this]() { m_presenter->setDirectoryChanged(true); });
+
+  // Timer to send timeout() signal intermitently
+  connect(m_timer, SIGNAL(timeout()), this, SLOT(notifyTimerEvent()));
```
There is a new signal-slot syntax which is now preferred in C++ (although its barely used in Mantid). The new syntax allows you to get compile-time warnings if there is a slot or signal that is missing. Previously, this would go undetected unless you keep an eye on the command prompt when opening your Qt interface
https://wiki.qt.io/New_Signal_Slot_Syntax

GuiMacielPereira replied:
Ah, so that's the difference! I encountered this style before but because the syntax looked a bit clunky to me at the time I just assumed that it was an older version of making connections. If I had looked this up it would have saved me some time during development :laughing:

robertapplin commented on qt/scientific_interfaces/Muon/ALCDataLoadingModel.cpp:
```diff
@@ -0,0 +1,346 @@
+ * This timer runs every second when we are watching a directory.
+ * If any changes have occurred in the meantime, reload.
+ */
+bool ALCDataLoadingModel::loadFilesFromWatchingDirectory(const std::string &firstFile,
```
This function is quite large with a lot of nested if statements. Would it be possible to refactor it to use the "return early" coding style? https://dev.to/jpswade/return-early-12o5#:~:text=The%20motivation%20behind%20this%20is,to%20follow%20the%20code%20low.

I find this improves code readability hugely!

robertapplin commented on qt/scientific_interfaces/Muon/ALCDataLoadingModel.cpp:
```diff
@@ -0,0 +1,346 @@
+ * If custom grouping is supplied, check all detector numbers are valid
+ * @returns :: True if grouping OK, false if bad
+ */
+bool ALCDataLoadingModel::checkCustomGrouping(const std::string &detGroupingType, const std::string &forwardGrouping,
```
Can we use the "return early" style in this function too?

robertapplin commented on qt/scientific_interfaces/Muon/ALCDataLoadingModel.cpp:
```diff
@@ -0,0 +1,346 @@
+
+  // Execute async so we can show progress bar
+  Poco::ActiveResult<bool> result(alg->executeAsync());
+  while (!result.available()) {
+    QCoreApplication::processEvents();
+  }
```
**No need to make any changes with regards to this comment**

This is something to be refactored/removed in a future PR. I'm sorry you had to come across this! It is trying to keep the GUI responsive while the algorithm is executed in a background thread, and whilst waiting for the algorithm to finish. It is much preferred to have a "callback" function that gets called when an algorithm is completed. This is functionality provided by the `AlgorithmRunner` class. This change can be complicated so best not to make any changes as part of this PR

GuiMacielPereira replied:
Sounds good!

robertapplin reviewed (changes requested):
Thank you for all your hard work with this! The code is much much improved! I had a few small comments, then I'm happy to approve. Also I still need to manually test

GuiMacielPereira committed:
Put function where it's used

GuiMacielPereira committed:
Make abstract class NOT inherit from QObject

GuiMacielPereira committed:
Set non-owning pointer for model inside test

Useful in case it gets used in the future

GuiMacielPereira committed:
Add const in several functions

Make getters const

GuiMacielPereira committed:
Refactor checkCustomGrouping into return early style

Retain the same functunality but improve readability by returning early.

GuiMacielPereira committed:
Refactor loading files into return early style

Made function easier to read.

GuiMacielPereira committed:
Update connections style

Using the new style allows for the compiler to detect missing connections during compilation,
rather than at run time.

GuiMacielPereira commented on qt/scientific_interfaces/Muon/IALCDataLoadingView.h:
```diff
@@ -7,22 +7,26 @@
  */
+class IALCDataLoadingPresenter;
+
 class MANTIDQT_MUONINTERFACE_DLL IALCDataLoadingView : public QObject {
```
Ah well spotted, I totally had missed it!

robertapplin reviewed (dismissed):
Thanks for your hard work refactoring into a better representation of the MVP pattern! The data loading part of the interface is working as expected and the code is much improved!

GuiMacielPereira commented on qt/scientific_interfaces/Muon/ALCDataLoadingModel.cpp:
```diff
@@ -0,0 +1,358 @@
+ * Load new data and update the view accordingly
+ * @param files :: [input] range of files (user-specified or auto generated)
+ */
+void ALCDataLoadingModel::load(const IALCDataLoadingView *view) {
```
@robertapplin I just remembered that I forgot to ask about this line, here the view is passed to the function from the model so I can get the inputs from the view. I avoided this in other model methods but this one requires quite a few of inputs. I don't know if there is a better way of doing this?

robertapplin replied:
Whoops I must have missed this! It would be preferred to pass each argument into the model `load` function individually so that the Model doesn't know about the view. Here the Model is communicating directly with the View instead of via the presenter

An alternative might be to use a std::tuple multiple items of data from the view to the model

robertapplin replied:
Thanks for pointing this out, I think its best to sort this out before we merge

GuiMacielPereira replied:
Okay, nice!

GuiMacielPereira committed:
Call inputs as arguments instead of passing view

Instead of passing the view to the model, it's best to pass the view inputs as arguments,
so that the model never has access to the view.

GuiMacielPereira committed:
Rename loading function

Made it clearer that load function in presenter triggers loading of
files that are *already present in the model*, so that in order to load files they have to
be loaded into the model first.

GuiMacielPereira commented on qt/scientific_interfaces/Muon/ALCDataLoadingView.cpp:
```diff
@@ -421,4 +442,43 @@ std::string ALCDataLoadingView::getAlphaValue() const {
+
+QTimer *ALCDataLoadingView::getTimer() { return m_timer; }
+
+void ALCDataLoadingView::handleStartWatching(bool watch) {
+  if (watch) {
+    // Get path to watch and add to watcher
+    const auto path = getPath();
+    m_watcher->addPath(QString::fromStdString(path));
+    // start a timer that executes every second
+    getTimer()->start(1000);
+
+  } else {
+    // Check if watcher has a directory, then remove all
+    if (!m_watcher->directories().empty()) {
+      m_watcher->removePaths(m_watcher->directories());
+    }
+    // Stop timer
+    getTimer()->stop();
+
+    m_presenter->handleWatcherStopped();
+  }
+}
```
As a final question, this is the only sort of logic present in the view, but I put it here because both `m_wacher` and `m_timer` are QObjects

robertapplin replied:
I think that reason is fair enough, so I'm happy for it to stay as it is!

GuiMacielPereira committed:
Fix unit test

Since view inputs are called as arguments of load function,
neeeded to change the number of times one of the arguments was called.

GuiMacielPereira committed:
Correct comment

robertapplin reviewed (dismissed):
I've tested this again and its still working as expected - thanks for the changes!

rbauststfc commented on qt/scientific_interfaces/Muon/ALCDataLoadingModel.cpp:
```diff
@@ -0,0 +1,362 @@
+  bool groupingOK = true;
+
+  auto detectors = Mantid::Kernel::Strings::parseRange(isCustomGroupingValid(forwardGrouping, groupingOK));
+  const auto backward = Mantid::Kernel::Strings::parseRange(isCustomGroupingValid(backwardGrouping, groupingOK));
```
Am I right that the logic here means that the final value of the `groupingOK` variable is only actually being determined by the custom grouping of backward (i.e. by line 256)? The custom grouping of the detectors in the step above doesn't influence the final value of `groupingOK`?

I just wanted to double check that that makes sense for what this is trying to achieve? I'm not sure of the real world context of this, so sorry if I'm missing something obvious here!

GuiMacielPereira replied:
My goodness, that's really well spotted, I'll fix it

rbauststfc replied:
In fairness, I think this was how the logic was originally, so if you prefer we could open an issue to fix the bug as a separate PR and get this one merged in?

GuiMacielPereira replied:
No, it's okay :relieved:

GuiMacielPereira committed:
Add check for forward grouping
